autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/securestorage/SecureStorageDatabaseFactory.kt

```diff
@@ -18,46 +18,85 @@ package com.duckduckgo.autofill.impl.securestorage
 
 import android.content.Context
 import androidx.room.Room
+import com.duckduckgo.autofill.api.AutofillFeature
 import com.duckduckgo.di.scopes.AppScope
 import com.duckduckgo.securestorage.store.db.ALL_MIGRATIONS
 import com.duckduckgo.securestorage.store.db.SecureStorageDatabase
 import com.squareup.anvil.annotations.ContributesBinding
 import dagger.SingleInstanceIn
 import javax.inject.Inject
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
 import net.sqlcipher.database.SupportFactory
 
 interface SecureStorageDatabaseFactory {
-    fun getDatabase(): SecureStorageDatabase?
+    suspend fun getDatabase(): SecureStorageDatabase?
 }
 
 @SingleInstanceIn(AppScope::class)
 @ContributesBinding(AppScope::class)
 class RealSecureStorageDatabaseFactory @Inject constructor(
     private val context: Context,
     private val keyProvider: SecureStorageKeyProvider,
+    private val autofillFeature: AutofillFeature,
 ) : SecureStorageDatabaseFactory {
     private var _database: SecureStorageDatabase? = null
 
+    private val mutex = Mutex()
+
+    override suspend fun getDatabase(): SecureStorageDatabase? {
+        // If we have already the DB instance then let's use it
+        return if (autofillFeature.canCreateAsyncPreferences().isEnabled()) {
+            getAsyncDatabase()
+        } else {
+            getDatabaseSynchronized()
+        }
+    }
+
     @Synchronized
-    override fun getDatabase(): SecureStorageDatabase? {
+    private fun getDatabaseSynchronized(): SecureStorageDatabase? {
         // If we have already the DB instance then let's use it
         if (_database != null) {
             return _database
         }
 
         // If we can't access the keystore, it means that L1Key will be null. We don't want to encrypt the db with a null key.
-        return if (keyProvider.canAccessKeyStore()) {
+        return if (keyProvider.canAccessKeyStoreSync()) {
```

Comment:
While technically the database will only be created once ans we should never be in a scenario where Key Store is created async and the database is created sync, I still don't love forcing an access to the sync version here.